### PR TITLE
Replace Engineering Diffraction Calibration and Focussing Python workflows with C++ workflows

### LINKED ISSUE #30048: Replace EnggCalibrate/Full algorithms with PDCalibration
PDCalibration is an existing C++ implementation that is widely used (and well tested) that performs a large chunk of the Calibration currently done in python in the EnggCalibrate and EnggCalibrateFull algorithms. The output of PDCalibrate is also compatible with other algorithms for unit conversion which should replace usages of ConvertUnits in the Engineering UI code.

### LINKED ISSUE #30288: Replace focussing in EngDiff UI with existing C++ algorithms and support user-defined grouping
The EngDiff UI focussing is currently done in python, yet all the necessary vanadium corrections and focussing can use existing faster and better tested C++ algorithms. In addition, it is necessary to support arbitrary grouping defined in a user-supplied .cal file, as well the option to focus each bank.

There are several smaller changes that would be good to make at the same time:
- Update instrument geometry with a single detector at the average calibrated two theta and L2 consistent with DIFC etc. to allow ConvertUnits to be used.
- Identify the bank/grouping by detector ID (necessary for use in rest of UI)
- Output focussed data in TOF and d-spacing

### PR DESCRIPTION
**Pull request by clayton-tom**
**Description of work.**

Calibration and Focussing (in both the gui and EnginX scripts) made use of slow and unreliable python algorithms. This PR replaces those workflows with CPP algorithm based workflows (PDCalibration).
Some other miscellaneous changes: also output focussed runs in dSpacing as well as TOF

**To test:**

1. Using the Engineering Diffraction gui, create a calibration (e.g vanadium 236516 ceria 193749)
2. Also create a calibration for a single bank, or for a crop of spectrum numbers (try 34-678 or something)
3. Focus a sample run using that calibration (e.g. 299080)
4. Check outputs on bank calibration to see if the peaks have a good agreement (fairly well aligned by dSpacing)
5. Give the EnginX script example a run with the same run number inputs and see if the outputs align between this and the gui

```from Engineering.EnginX import main

path = '/change/this/path/to/whole_inst_calib.nxs'
main(vanadium_run='236516', user='0110', focus_run='299080', ceria_run='193749', full_inst_calib_path=path)

# Kwargs:
#         ceria_run (string): the run number of the ceria to use
#         force_vanadium (bool): forces creation of vanadium even if one for the run already exists
#         force_cal (bool) : forces creation of calibration files
#         full_inst_calib_path (string) : path to the full instrument calibration file used in calibration and focussing
#         crop_type (string): what method of cropping to use
#         crop_name (string): what to call the cropped bank workspace
#         crop_on (string): the bank of spectrum to crop on if cropping
#         pre_process_run (bool): set whether or not to pre-process run before focusing
#         params (string): rebin parameters for pre-process
#         time_period (string): time period for pre-process
#         grouping_file (string): the path of the grouping file for texture focusing
#         directory (string): the path of the directory to save to
#         user_struct (bool): whether or not to use the enginx file structure (defaults to true)
```

Full instrument calibration file (needed for both the gui and scripts):
[full_inst_calib.zip](https://github.com/mantidproject/mantid/files/6473094/full_inst_calib.zip)

Fixes #30048 and Fixes #30288 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
I think something has gone wrong with the whole instrument calibration included in the instructions - I re-ran using my original calibration script (which can be found here [full_inst_calib_RWaite.zip](https://github.com/mantidproject/mantid/files/6538654/full_inst_calib_RWaite.zip)) and got much better results (closer agreement of focussed spectra in each bank and better d-spacing of peaks)
image
Is it possible to package the whole instrument calibration as e.g. a csv or nexus with mantid?

**Comment by clayton-tom** [comment]
> I think something has gone wrong with the whole instrument calibration included in the instructions - I re-ran using the cal workspace from my original calibration script (which can be found here [full_inst_calib_RWaite.zip](https://github.com/mantidproject/mantid/files/6538654/full_inst_calib_RWaite.zip)) and got much better results (closer agreement of focussed spectra in each bank and better d-spacing of peaks)
> image
> Is it possible to package the whole instrument calibration as e.g. a csv or nexus with mantid?

I had a chat with Danny about this just now, he's happy for us to put it in `scripts/Engineering/calib` as this seems to be where the old csv calibration lived, and gets shipped with installs

**Comment by RichardWaiteSTFC** [comment]
> I had a chat with Danny about this just now, he's happy for us to put it in `scripts/Engineering/calib` as this seems to be where the old csv calibration lived, and gets shipped with installs

Great thanks for that, is it possible to set this as the default full instrument calibration (rather than have none)

File: scripts/Engineering/gui/engineering_diffraction/settings/settings_presenter.py [hunk 1]
```diff
@@ -5,14 +5,16 @@
 #   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
 # SPDX - License - Identifier: GPL - 3.0 +
 # pylint: disable=invalid-name
-from os import path
+from os import path, getcwd
 from mantidqt.utils.observer_pattern import Observable
 
+CALIB_FOLDER = path.join(path.dirname(path.dirname(path.dirname(getcwd()))), "calib")
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/settings/settings_presenter.py:11]: Is there a better way than calling `dirname` three times (I don't know if there is!). I was wondnering if it would automatically be on the path (in the same way the Parameters.xml files are) and we can use the same filefinder stuff we use in algorithms?

Comment (clayton-tom) [reply scripts/Engineering/gui/engineering_diffraction/settings/settings_presenter.py:11]: I'll have a look, might be possible to add the calib folder to the default data search directories

File: scripts/Engineering/EnginX.py [hunk 2]
```diff
@@ -381,17 +582,22 @@ def create_difc_zero_workspace(difc, tzero, crop_on, name):
         actual_i = correction + i
         # retrieve required workspace
         if not plot_spec_num:
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:584]: Can you think of a nicer more readable way to select which banks to retrieve  - I don't like L569-578. I know that it allows you to put this code in a loop (and I would want to keep the loop) but it seems like a lot of work and I think the `correction` variable is quite confusing (it adjusts the bank number to be North or South, but perhaps it's nicer just populate the `banks` list correctly in the first place - i.e. `banks = [1]` for North and `banks = [2]` for South).

File: scripts/Engineering/EnginX.py [hunk 3]
```diff
@@ -381,17 +582,22 @@ def create_difc_zero_workspace(difc, tzero, crop_on, name):
...
         x_val = []
         y_val = []
         y2_val = []
-        for irow in range(0, bank_ws.rowCount()):
-            x_val.append(bank_ws.cell(irow, 0))
-            y_val.append(bank_ws.cell(irow, 5))
+        for irow in range(0, fitparam_ws.rowCount()):
+            x_val.append(expected_dspacing_peaks[-(irow+1)])
+            y_val.append(fitparam_ws.cell(irow, 5))
             y2_val.append(x_val[irow] * difc[i - 1] + tzero[i - 1])
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:601]: 1. Why is there no `difa` here and what exactly is `y2` - is it meant to be the fitted quadratic curve to TOF(d)? If so perhaps we call this `yfit`? Generally speaking I don't like the variables names `x_val`, `y_val `etc. (I know they pre-date your changes but since you're familiar with their purpose it would make sense for you to change them in this PR).
2. I think there is a unit converter exposed to python that will convert single values (not workspaces) based of the diffractometer constants so we don't have to reproduce this equation each time. I'm having trouble finding it, @DannyHindson do you know?
3. Also it would be good to retrieve the errors on the TOF as well to plot.

File: scripts/Engineering/EnginX.py [hunk 4]
```diff
@@ -175,125 +220,268 @@ def create_calibration(ceria_run, van_run, calibration_directory, calibration_ge
...
+        "CalibrationParameters": 'DIFC+TZERO+DIFA',
+        "UseChiSq": True
+    }
+    cal_output = dict()
+    curves_output = list()
+
+    if spectrum_numbers is None:
+        if bank == '1' or bank is None:
+            df_kwarg = {"GroupingFileName": NORTH_BANK_CAL}
+            ws_d_clone = simple.CloneWorkspace(ws_d)
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:382]: I might be missing something but I think we don't need to clone these workspaces before focussing?

File: scripts/Engineering/EnginX.py [hunk 5]
```diff
@@ -578,14 +838,22 @@ def focus_texture_mode(run_number, van_curves, van_int, focus_directory, focus_g
...
-    # loop through the banks described in the csv, focusing and saing them out
+    # loop through the banks described in the csv, focusing and saving them out
     for bank in banks:
-        output_ws = "engg_focusing_output_ws_texture_bank_{}"
-        output_ws = output_ws.format(bank)
-        simple.EnggFocus(InputWorkspace=ws_to_focus, OutputWorkspace=output_ws,
-                         VanIntegrationWorkspace=van_integrated_ws, VanCurvesWorkspace=van_curves_ws,
-                         SpectrumNumbers=banks.get(bank))
-        _save_out(run_number, focus_directory, focus_general, output_ws, "ENGINX_{}_texture_{}{{}}", bank)
+        sample_ws_clone = simple.CloneWorkspace(ws_to_focus)
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:843]: Do we need to clone these workspaces if we apply the sensitivity correction outside the loop over banks?

File: scripts/Engineering/EnginX.py [hunk 6]
```diff
@@ -175,125 +220,268 @@ def create_calibration(ceria_run, van_run, calibration_directory, calibration_ge
...
+    ws_van_d = simple.ConvertUnits(InputWorkspace=ws_van, Target='dSpacing')
+
+    ws_van_d /= van_integration
+
+    # sensitivity correction for sample
+    simple.NormaliseByCurrent(InputWorkspace=sample_ws, OutputWorkspace=sample_ws)
+    simple.ApplyDiffCal(InstrumentWorkspace=sample_ws, CalibrationWorkspace=full_inst_calib)
+    ws_d = simple.ConvertUnits(InputWorkspace=sample_ws, Target='dSpacing')
+
+    ws_d /= van_integration
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:362]: We don't really need to do any form of vanadium correction (even sensitivity) to the ceria run if we only use it for PDCalibration (of focussed or whole instrument) - the vanadium corrections only effect the intensities.

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Great effort, this turned into a large issue! I have a few comments, mostly on the low-level functions, I have yet to review the UI stuff.

File: scripts/Engineering/EnginX.py (hunk 3 above)
Comment (RichardWaiteSTFC) [reply scripts/Engineering/EnginX.py:601]: Ah here I think is how to use the UnitConverter in python
https://github.com/mantidproject/mantid/blob/aca78985607ac37e8a57ef3f16774e891b9b106d/scripts/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py#L340

**Comment by RichardWaiteSTFC** [comment]
Looks like typo in an assertion in `python.EngineeringDiffraction.test_focus_model.test_focus_model`
`'305761_engggui_focusing_output_ws_bank_cropped'` should be `'305761_engggui_focusing_output_ws_bank_Cropped'` (i.e. capital C on cropped)

**Comment by RichardWaiteSTFC** [comment]
In #31213 I added a check to see if proton charge was finite before calling Normalise by current - I think we might need a similar check in this new workflow (though I haven't tested it with zero proton charge data yet).

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py [hunk 7]
```diff
@@ -36,77 +42,107 @@ def focus_run(self, sample_paths, banks, plot_output, instrument, rb_num, spectr
...
-            return EnggFocus(**kwargs)
-        except RuntimeError as e:
-            logger.error(
-                "Error in focusing, Could not run the EnggFocus algorithm successfully for bank "
-                + str(bank) + ". Error Description: " + str(e))
-            raise RuntimeError()
+                   df_kwarg,
+                   full_calib,
+                   region_calib):
+        NormaliseByCurrent(InputWorkspace=input_workspace, OutputWorkspace=input_workspace)
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:128]: In #31213 I added a check to see if proton charge was finite before calling Normalise by current - I think we might need a similar check in this new workflow (though I haven't tested it with zero proton charge data yet).

File: scripts/Engineering/EnginX.py [hunk 8]
```diff
@@ -14,11 +14,16 @@
 import Engineering.EnggUtils as Utils
 import mantid.simpleapi as simple
 
+NORTH_BANK_CAL = "EnginX_NorthBank.cal"
+SOUTH_BANK_CAL = "EnginX_SouthBank.cal"
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:18]: Maybe this is for another PR - but I have had a go at using the algorithm CreateGroupingWorkspaceand eventually figured out how to get it working! So instead of keeping these cal files (one for each bank) we can instead get a grouping workspace to focus e.g. for the north bank
```
Load(Filename=r'\\isis.cclrc.ac.uk\inst$\ndxenginx\instrument\data\cycle_12_4\ENGINX00193749.nxs', OutputWorkspace='193749')
CreateGroupingWorkspace(InputWorkspace='193749', GroupNames='NorthBank', OutputWorkspace='north_grp')
DiffractionFocussing(InputWorkspace='193749', OutputWorkspace='foc', GroupingWorkspace='north_grp')

```

**Comment by RichardWaiteSTFC** [comment]
I pulled the latest changes and am now getting an error on the focus tab (after doing a fresh calibration with no cropping)
```
Exception in thread Thread-11:
Traceback (most recent call last):
  File "C:\mantid\external\src\ThirdParty\lib\python3.8\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\mantid\qt\python\mantidqt\utils\asynchronous.py", line 78, in run
    self.finished_cb()
  File "C:\mantid\scripts\Engineering\gui\engineering_diffraction\tabs\focus\presenter.py", line 74, in _on_worker_success
    self.focus_run_notifier.notify_subscribers(self.model.get_last_path())
  File "C:\mantid\qt\python\mantidqt\utils\observer_pattern.py", line 99, in notify_subscribers
    Observable.notify_subscribers(self, *args)
  File "C:\mantid\qt\python\mantidqt\utils\observer_pattern.py", line 55, in notify_subscribers
    self.thread_safe_update_call(arg, **kwargs)
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 52, in __call__
    raise self._exc_info[1].with_traceback(self._exc_info[2])
  File "C:\mantid\qt\python\mantidqt\utils\qt\qappthreadcall.py", line 62, in on_call
    self.callee(*self._args, **self._kwargs)
  File "C:\mantid\qt\python\mantidqt\utils\observer_pattern.py", line 59, in _notify_subscribers_impl
    observer.update(self, arg, **kwargs)
  File "C:\mantid\qt\python\mantidqt\utils\observer_pattern.py", line 91, in update
    self.callback(arg)
  File "C:\mantid\scripts\Engineering\gui\engineering_diffraction\tabs\fitting\data_handling\data_view.py", line 70, in set_file_last
    directory, discard = path.split(filepath)
  File "C:/mantid/external/src/ThirdParty\lib\python3.8\lib\ntpath.py", line 177, in split
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType

```

File: scripts/Engineering/gui/engineering_diffraction/settings/settings_presenter.py (hunk 1 above)
Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/settings/settings_presenter.py:11]: did you find anything to help with this?

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py [hunk 9]
```diff
@@ -93,39 +97,43 @@ def create_new_calibration(self,
...
+        for param_name in ["alpha", "beta_0", "beta_1", "sigma_0_sq", "sigma_1_sq", "sigma_2_sq"]:
             params_north += [NorthBank.getNumberParameter(param_name)[0]]
             params_south += [SouthBank.getNumberParameter(param_name)[0]]
 
-        return [params_north,params_south]
+        return [params_north, params_south]
 
-    def load_existing_gsas_parameters(self, file_path):
+    def load_existing_calibration_files(self, file_path):
+        # TODO write test this function in conjunction with get_info_from_file once neat way of getting test file found
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:123]: Is this comment still relevant?

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py [hunk 10]
```diff
@@ -146,144 +154,167 @@ def update_calibration_params_table(params_table):
...
+        plot_res_2 = fig.add_subplot(gs[1, 1], projection="mantid")
 
         for ax, ws, bank in zip([plot_bank_1, plot_bank_2], [bank_1_ws, bank_2_ws], [1, 2]):
             self._add_plot_to_axes(ax, ws, bank)
+        for ax, ws in zip([plot_res_1, plot_res_2], [bank_1_residuals_ws, bank_2_residuals_ws]):
+            self._add_residuals_to_axes(ax, ws)
+        fig.tight_layout()
         fig.show()
 
     def _plot_tof_fit_single_bank_or_custom(self, bank):
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:177]: Is there not a way of combining this function with `_plot_tof_fit`? Once you have established how many banks (i.e. how many columns of subplots) should this not just be a loop over nbanks (which may be 1)?

**Comment by RichardWaiteSTFC** [comment]
Ah I think something has gone wrong with your merge - I now get ~400 commits and files changed!

File: scripts/Engineering/EnginX.py (hunk 8 above)
Comment (clayton-tom) [reply scripts/Engineering/EnginX.py:18]: Thanks for investigating that, I'll make that part of the custom cal file PR

File: scripts/Engineering/EnginX.py [hunk 11]
```diff
@@ -111,57 +128,86 @@ def run(ceria_run, do_cal, do_van, van_run, calibration_directory, calibration_g
...
+    # make the integration workspace
+    simple.NormaliseByCurrent(InputWorkspace=van_ws, OutputWorkspace=van_ws)
+    # sensitivity correction for van
+    ws_van_int = simple.Integration(InputWorkspace=van_ws)
+    ws_van_int /= van_ws.blocksize()
     # save out the vanadium and delete the original workspace
-    simple.SaveNexus("eng_vanadium_integration", van_int_file)
-    simple.SaveNexus("eng_vanadium_curves", van_curves_file)
+    simple.SaveNexus(ws_van_int, van_int_file)
     simple.DeleteWorkspace(van_name)
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/EnginX.py:187]: Very minor, but the algorithm `DeleteWorkspaces` can delete a list of workspaces - I've seen repeated calls to `DeleteWorkspace` in a  few places but it's not a big deal.

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py [hunk 12]
```diff
@@ -146,144 +154,167 @@ def update_calibration_params_table(params_table):
...
+        ax.errorbar(ws, color='b', marker='.', wkspIndex=0, capsize=2, ls="None")
+        ax.axhline(color='r')
         ax.set_xlabel("Expected Peaks Centre(dSpacing, A)")
-        ax.set_ylabel("Fitted Peaks Centre(TOF, us)")
+        ax.set_ylabel("Residuals \u03BCs")
 
-    def run_calibration(self,
-                        sample_ws,
+    @staticmethod
+    def run_calibration(ceria_ws,
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:207]: As discussed offline this code is repeated in the EnginX script - I assume the duplicated code will eventually be put in something like EnggUtils in PR #31686

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py [hunk 13]
```diff
@@ -36,77 +42,114 @@ def focus_run(self, sample_paths, banks, plot_output, instrument, rb_num, spectr
...
+            logger.warning(f"Skipping focus of run {input_workspace.name()} because it has invalid proton charge.")
+            return False
+        input_workspace /= vanadium_integration_ws
+        ReplaceSpecialValues(InputWorkspace=input_workspace, OutputWorkspace=input_workspace, NaNValue=0,
+                             InfinityValue=0)
+        ApplyDiffCal(InstrumentWorkspace=input_workspace, CalibrationWorkspace=full_calib)
+        ws_d = ConvertUnits(InputWorkspace=input_workspace, Target='dSpacing')
+        focused_sample = DiffractionFocussing(InputWorkspace=ws_d, **df_kwarg)
+        curves_rebinned = RebinToWorkspace(WorkspaceToRebin=vanadium_curves_ws, WorkspaceToMatch=focused_sample)
+        normalised = Divide(LHSWorkspace=focused_sample, RHSWorkspace=curves_rebinned,
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:142]: We can avoid making another workspace here if you overwrite `focused_sample` - you should also be able to make use of overridden operators e.g. `focused_sample /= curves_rebinned` - I think this will still show up in the workspace history as `Divide`?

Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:142]: Similarly it should also be OK to overwrite the vanadium curves with the rebinned curves?

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py [hunk 14]
```diff
@@ -36,77 +42,114 @@ def focus_run(self, sample_paths, banks, plot_output, instrument, rb_num, spectr
...
+                            AllowDifferentNumberSpectra=True)
+        ApplyDiffCal(InstrumentWorkspace=normalised, CalibrationWorkspace=region_calib)
+        dspacing_output_name = tof_output_name + "_dSpacing"
+        CloneWorkspace(InputWorkspace=normalised, OutputWorkspace=dspacing_output_name)
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:146]: Is this not actually intended to be a Rename rather than Clone - we may nbot need either if we call it by the correct name above by specifying `OutputWorkspace` in `DiffractionFocusing `

Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:146]: There are maybe a few more places we could cut down on the number of workspaces by specifying the OutputWorkspace - I will have another look.

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py (hunk 13 above)
Comment (clayton-tom) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:142]: We need to keep the `AllowDifrerentNumberSpectra` kwarg which I don't think we can replicate with the `/` operator

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py (hunk 12 above)
Comment (clayton-tom) [reply scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:194]: Yep, plan is to sort that out as part of that more general PR

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py (hunk 13 above)
Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:142]: Hmm, I thought the `/=` was clever enough to do the required broadcasting - but I haven't tried it!

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py (hunk 14 above)
Comment (clayton-tom) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:146]: Did you mean ApplyDiffCal rather than DiffractionFocusing? if so that's an in place op so doesn't take an output workspace, but I think we can do what you had in mind by just renaming `normalised`

Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py]: I had a separate comment above related to the divide being in-place such that `normalised` is still `focused_sample` - so here I am saying can we call `focused_sample` (output of a call to DiffractionFocusing) whatever dspacing_output_name is (i.e. we have OutputWorkspace=tof_output_name+"_dSpacing" on L140 of this file?

**Comment by RichardWaiteSTFC** [comment]
Just a couple more things I've noticed.
(1) One of the calibration tab output workspaces is `engggui_vanadium_ws` which is the raw vanadium in TOF - do we need to keep this around?
(2) It appears that the diff constants for bank 2 are not cinsistent bewtween tables (aside from the inconsistency once we have got the summary of both banks in a table we can probably delete the cal form the PDCalibration for each bank)
image

From the focussed d-spacing data it looks like the values from the summary has been used
image

**Comment by RichardWaiteSTFC** [comment]
Also when I load in a calibration it loads `engggui_vanadium_ws` (the whole instrument vanadium data) not the focussed curves (which I was expecting) - when I go to focus i get an error saying load a calibration - perhaps the two are related?

**Comment by RichardWaiteSTFC** [comment]
1. So when I perform a calibration which is cropped to use only the North Bank and then focus I get 3 workspaces: the North bank in TOF and both banks in dSpacing - except I don't think they are the two banks as they look exactly the same. 
2. I'm a bit confused because you can calibrate the South Bank only and then in the focussing tab crop to North Bank (presumably using the vanadium curves from the South Bank?)- surely the crop option in the focus tab depends on what calibration workspaces exist/what the previous calibration was?
3. Also can we remove a few more workspaces post-calibration - I think it's not necessary to have the `engggui_calibration_bank_` as these are summarised in a different table with 1 row/bank. And also remove the workspace for the tof(d) and residuals - actually we could avoid making workspaces at all they're only used for plotting.

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py [hunk 15]
```diff
@@ -132,19 +131,25 @@ def _run_focus(input_workspace,
...
-        ApplyDiffCal(InstrumentWorkspace=normalised, CalibrationWorkspace=region_calib)
+        ConvertUnits(InputWorkspace=input_workspace, OutputWorkspace=input_workspace, Target='dSpacing')
+        # rename workspace prior to focussing to avoid errors later
         dspacing_output_name = tof_output_name + "_dSpacing"
-        RenameWorkspace(InputWorkspace=normalised, OutputWorkspace=dspacing_output_name)
-        ConvertUnits(InputWorkspace=normalised, OutputWorkspace=tof_output_name, Target='TOF')
-        DeleteWorkspaces([curves_rebinned, focused_sample, ws_d])
+        RenameWorkspace(input_workspace, dspacing_output_name)
+        # focus over specified region of interest
+        DiffractionFocussing(InputWorkspace=input_workspace, OutputWorkspace=dspacing_output_name,
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:143]: I think this call to `DiffractionFocusing` overwrites the whole instrument data - as discussed offline you clone the instrument before this function so the overwriting is OK, but it is a bit confusing and ideally we want to avoid cloning the whole instrument - to this end can we put the sensitivity and unit conversion stuff outside the loop over banks?

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py [hunk 16]
```diff
@@ -99,6 +104,11 @@ def focus_run(self, sample_paths, banks, plot_output, instrument, rb_num, spectr
                     else:
                         df_kwarg = {"GroupingFileName": SOUTH_BANK_CAL}
                         region_calib = "engggui_calibration_bank_2"
+                    # check correct region calibration exists
+                    if not Ads.doesExist(region_calib):
```
Comment (RichardWaiteSTFC) [inline scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:121]: This fixes the problem I saw but can this check go earlier? No worries if not/it is a lot of effort- I know we plan to consolidate the ROI selection into a single place in the UI so it is fine for now.

**Comment by RichardWaiteSTFC** [comment]
This works well - the calibration looks much better, the graphs shown are much more informative and your refactoring to support the introduction of cal/custom ROI. When you fix the broken test I'll be happy to approve (there's a small comment above but I suspect that code is due to change imminently anyway  - if so free free to leave it).

File: scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py (hunk 16 above)
Comment (clayton-tom) [reply scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py:121]: I think to put it earlier we'd have to duplicate the check to see what region of interest is selected, and to time of execution for the code before it returns is fairly small, so in my opinion I think this is best left for the PR to include the single ROI selection

**Review by RichardWaiteSTFC — COMMENTED** [review]
Oh no, just ran the EnginX script and get this error that I think is from one of your recent changes!
```
Traceback (most recent call last):
  File "C:/mantid/scripts/Engineering/EnginX.py", line 917, in <module>
    main(vanadium_run='236516', user='0110', focus_run='299080', ceria_run='193749', full_inst_calib_path=path)
  File "C:/mantid/scripts/Engineering/EnginX.py", line 97, in main
    run(ceria_run, force_cal, force_van, full_inst_calib, vanadium_run, calibration_directory, calibration_general, cropped,
  File "C:/mantid/scripts/Engineering/EnginX.py", line 151, in run
    create_calibration(ceria_run, van_run, full_inst_calib, calibration_directory, calibration_general, cropped,
  File "C:/mantid/scripts/Engineering/EnginX.py", line 230, in create_calibration
    create_calibration_files(ceria_run, van_run, full_inst_calib, van_int_file, van_curves_file,
  File "C:/mantid/scripts/Engineering/EnginX.py", line 304, in create_calibration_files
    Utils.generate_tof_fit_dictionary(f"bank_{i}", "", "engg_tof_peaks_")
TypeError: generate_tof_fit_dictionary() takes from 1 to 2 positional arguments but 3 were given
```

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks good this is a big improvement. Great work on getting this over the line. There are a few bits and pieces (discussed in comments above) which can be dealt with in a later PR that separates the vanadium calibration from the ceria and consolidates so that the user only has to select to ROI once.

**Comment by DannyHindson** [comment]
@RichardWaiteSTFC @clayton-tom - well done both on getting this PR through
